bucket-scanner/pkg/counter/replication/counter_set.go

```diff
@@ -0,0 +1,119 @@
+package replication
+
+import (
+	"sync"
+
+	"github.com/scality/backbeat/bucket-scanner/pkg/types"
+)
+
+type (
+	CounterSet struct {
+		sync.Mutex
+		Global           *ByStatusCounterSet
+		ByTargetLocation map[types.SiteName]*ByStatusCounterSet
+
+		name string
+	}
+
+	SerializedCounterSet struct {
+		Global     *SerializedByStatusCounterSet                    `json:"global"`
+		ByLocation map[types.SiteName]*SerializedByStatusCounterSet `json:"byLocation"`
+	}
+)
+
+func New(name string) *CounterSet {
+	return &CounterSet{
+		Global:           NewByStatusCounterSet(name + ":global"),
+		ByTargetLocation: make(map[types.SiteName]*ByStatusCounterSet),
+
+		name: name,
+	}
+}
+
+func (cs *CounterSet) CopyTo(other *CounterSet) {
+	other.Lock()
+	defer other.Unlock()
+
+	cs.Global.CopyTo(other.Global)
+
+	other.ByTargetLocation = make(map[types.SiteName]*ByStatusCounterSet)
+	for k, v := range cs.ByTargetLocation {
+		n := NewByStatusCounterSet(other.name + ":" + string(k))
+		v.CopyTo(n)
+		other.ByTargetLocation[k] = n
+	}
+}
+
+func (cs *CounterSet) Add(other *CounterSet) {
+	cs.Lock()
+	defer cs.Unlock()
+
+	cs.Global.Add(other.Global)
+
+	for k, v := range other.ByTargetLocation {
+		m := cs.getTargetLocation(k)
+		m.Add(v)
+	}
+}
+
+func (cs *CounterSet) Reset() {
+	cs.Lock()
+	defer cs.Unlock()
+
+	cs.Global.Reset()
+	cs.ByTargetLocation = make(map[types.SiteName]*ByStatusCounterSet)
+}
+
+func (cs *CounterSet) GetTargetLocation(location types.SiteName) *ByStatusCounterSet {
+	cs.Lock()
+	defer cs.Unlock()
+
+	return cs.getTargetLocation(location)
+}
+
+func (cs *CounterSet) getTargetLocation(location types.SiteName) *ByStatusCounterSet {
+	bscs, ok := cs.ByTargetLocation[location]
+	if !ok {
+		bscs = NewByStatusCounterSet(cs.name + ":" + string(location))
+		cs.ByTargetLocation[location] = bscs
+	}
+
+	return bscs
+}
+
+func (cs *CounterSet) Serialize() *SerializedCounterSet {
+	cs.Lock()
+	defer cs.Unlock()
+
+	byLocation := map[types.SiteName]*SerializedByStatusCounterSet{}
+
+	for location, bscs := range cs.ByTargetLocation {
+		byLocation[location] = bscs.Serialize()
+	}
+
+	return &SerializedCounterSet{
+		Global:     cs.Global.Serialize(),
+		ByLocation: byLocation,
+	}
+}
+
+func (cs *CounterSet) LoadSerialized(s *SerializedCounterSet) {
+	cs.Global.LoadSerialized(s.Global)
+
+	for k, v := range s.ByLocation {
+		l := cs.GetTargetLocation(k)
+		l.LoadSerialized(v)
+	}
+}
```

Comment:
please correct me if I am wrong.

With the lock being claimed/released per location as opposed on the method, it would seem like a race condition may occur if the `LoadSerialized` call is followed by a write method like `Add`, leading to an unexpected final state.

A read method after a `LoadSerialized` can also result in an invalid output as well?

Reply:
I don't think it's an issue because:

- It's only used to deserialize values from storage at init time, before the `CounterSet` gets passed around.
- Actual modification is done in a thread-safe manner using atomic ops, so the risk is that global counters would be slightly off compared to the sum of the regions on a functional level, not actual unsafe memory access that could be seen/reported by the race detector. Since this whole counting system is eventually consistent and there can be some known (temporary) double-counting happening, no one would notice :)
- The `Add` method should, if that was a use case to support, use atomic addition ops instead of using `Set(Get()+Get())`, making the pattern you describe safe.

But thanks for noticing, it's a valid concern.